### PR TITLE
Add support of contracts-v4 changes

### PR DESCRIPTION
**Pull request by MaximFischuk**

### CODE REVIEW

File: src/main/java/io/zksync/signer/ZkSigner.java [hunk 1]
```diff
@@ -115,6 +116,8 @@ public ChangePubKey signChangePubKey(ChangePubKey changePubKey) {
             outputStream.write(tokenIdToBytes(changePubKey.getFeeToken()));
             outputStream.write(feeToBytes(changePubKey.getFeeInteger()));
             outputStream.write(nonceToBytes(changePubKey.getNonce()));
+            outputStream.write(numberToBytesBE(changePubKey.getTimeRange().getValidFrom(), 8));
```
Comment (StanislavBreadless) [inline src/main/java/io/zksync/signer/ZkSigner.java:119]: In the JS SDK the timestamps are serialized the following way:

```
export function serializeTimestamp(time: number): Uint8Array {
    if (time < 0) {
        throw new Error('Negative timestamp');
    }
    return ethers.utils.concat([new Uint8Array(4), numberToBytesBE(time, 4)]);
}
```

I am not sure that the expression `numberToBytesBE(time, 8)` in your code is an equivalent one.

**Review by StanislavBreadless — COMMENTED** [review]
It seems to me that a few changes missing:

1. The `deposit` function still calls the old smart contract functions.
2. The fee for ChangePubKey should now be fetched with a little different params (look at the provider code in the JS SDK).

**Comment by StanislavBreadless** [comment]
Also for greater usability, it should be possible to create a default `TimeRange` with maximum possible `to` and minimum possible `from` and/or to be able to set its value as null in the params list and/or omitt it in params.

I am not sure what is more idiomatic in Java, but one way or another the user should be able to  do transfers without typing the big numbers by themselves.

File: src/main/java/io/zksync/domain/auth/ChangePubKeyCREATE2.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    private final ChangePubKeyAuthType type = ChangePubKeyAuthType.CREATE2;
+
+    private String creatorAddress;
+    private String saltArg;
+    private String codeHash;
+
+    @Override
+    public byte[] getBytes() {
+        return new byte[32];
```
[the remark this answers is not in the document]
Comment (MaximFischuk) [reply src/main/java/io/zksync/domain/auth/ChangePubKeyCREATE2.java:18]: Here is low cost abstraction for method that creates auth data of change pub key transaction
`public static byte[] getChangePubKeyData(String pubKeyHash, Integer nonce, Integer accountId, ChangePubKeyVariant changePubKeyVariant) {
        final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        outputStream.write(addressToBytes(pubKeyHash));
        outputStream.write(nonceToBytes(nonce));
        outputStream.write(accountIdToBytes(accountId));
        outputStream.write(changePubKeyVariant.getBytes()); // <===

        return outputStream.toByteArray();
    }`
As I understand v4 specification requires append 32 zeros to the end of an output message if it isn't batch.

> The deposit function still calls the old smart contract functions.

Can you explain what the function must be used? Cause I see that both [js](https://github.com/matter-labs/zksync/blob/5f47fe9990ec87e3087d32d083d13e6cab331ff1/sdk/zksync.js/src/wallet.ts#L803) and [rust](https://github.com/matter-labs/zksync/blob/5f47fe9990ec87e3087d32d083d13e6cab331ff1/sdk/zksync-rs/src/ethereum/mod.rs#L264) implementations still using

File: src/main/java/io/zksync/signer/ZkSigner.java (hunk 1 above)
Comment (MaximFischuk) [reply src/main/java/io/zksync/signer/ZkSigner.java:119]: Actually it's the same :) Just in JS uses integer 32 with leading zeros but in rust implementation it is u64. I've used the rust's one

File: src/main/java/io/zksync/domain/auth/ChangePubKeyCREATE2.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (StanislavBreadless) [reply src/main/java/io/zksync/domain/auth/ChangePubKeyCREATE2.java:18]: My bad. All good 👍

**Review by StanislavBreadless — APPROVED** [review]

**Pull request merged**